src/redisearch_rs/inverted_index/src/controlled_cursor.rs

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+use std::io::{IoSlice, Seek, SeekFrom, Write};
+
+/// Custom implementation of [`std::io::Cursor`] which uses a custom a conservative growth strategy
+/// when writing to the underlying vec. This is needed because the default Cursor uses a
+/// [`Vec::reserve`] call which uses a doubling strategy. This can lead to excessive memory usage
+/// for inverted index leafs with a small number of documents and therefore a small buffer inside
+/// [`crate::IndexBlock`]s. This is common for text indexes where many terms are rare and only appear in a
+/// few documents.
+///
+/// This implementation uses a [`Vec::reserve_exact`] call to only allocate the exact amount of memory
+/// needed to write the data. This can lead to more frequent allocations, but avoids the excessive
+/// memory usage caused by the doubling strategy.
+///
+/// The rest of this code is a verbatim copy of the [`std::io::Cursor`] implementation.
```

Comment:
It may be useful to use a clear comment tag in the code with the bits we changed to make syncing easier if we ever need to update the code from the std version.